pkg/cache/lru/lru.go

```diff
@@ -113,10 +113,10 @@ func (c *LRU[K, V]) removeOldest() {
 func (c *LRU[K, V]) removeElement(e *list.Element) {
 	c.evictList.Remove(e)
 	kv := e.Value.(entry[K, V])
-	delete(c.items, kv.key)
 	if c.onEvicted != nil {
 		c.onEvicted(kv.key, kv.value)
 	}
+	delete(c.items, kv.key)
```

Comment:
Sooooo if we delete the element *before* calling the callback it could happen that we attempt to close the file after it's been garbage collected. If it happens concurrently (there is a finalizer on files by default) then it might race and attempt to close a closed file.

By deleting it afterwards we would prevent this... the problem is... onEvicted might run in a goroutine and if that's the case changing this order won't do it, we will have to either not run the `onEvicted` to run serially or do something fancier @kakkoyun

Reply:
This is the issue that I think happens due to the above https://github.com/parca-dev/parca-agent/issues/1798

Reply:
Yes, this is a good catch. Let me think about this for a while.

Reply:
> Sooooo if we delete the element _before_ calling the callback it could happen that we attempt to close the file after it's been garbage collected.

Why do you think this is? Since we give a reference of the file to call the callback, GC wouldn't touch this file. Also if are we talking about this panic https://github.com/parca-dev/parca-agent/issues/1798 ? Because if it's the case it's a panic that we created https://github.com/parca-dev/parca-agent/blob/fee025041d725eb361487ec704dbd3eea80d265c/pkg/objectfile/object_file.go#L67-L72 So we try to read a file after we evict it from the cache.

Reply:
I think you are right:

```
kv := e.Value.(entry[K, V])
```

We get a reference to the value and then `onEvicted` and `delete` don't have to be serialized. I am assuming that the delete won't clear the backing memory for the entry, which I am not sure if it's true, though